chess-AI/mcts.py

```diff
@@ -12,147 +15,125 @@ class Mcts:
-
-    def find_search_probs(self, fen_string):
-        """Calculates and returns the search probabilites from the n_values for given fen_string.
-        """
-        # Calculates the search probabilites based on the the number of times a node has
-        # been visited/sum of all nodes visited
-        
+
+    def update_qn(self, fen_string, uci_move, value):
+        """Update q and n values after an expansion"""
+        n_value, q_value = self.state_values[fen_string][uci_move]
+
+        if not isinstance(value, int):
+            value = float(value.numpy())
+
+        # Average new value into q_value and increment n_value for the new node visit
+        self.state_values[fen_string][uci_move][1] = (n_value * q_value + value)/(n_value + 1)
+        self.state_values[fen_string][uci_move][0] += 1
+
+    def find_search_probs(self, fen_string, temperature):
+        """Returns the overall search probabilities after search() has been run"""
         # TODO: Create correct formula for calculating search probs
-        number_of_node_visits = np.array(list(self.n_values[fen_string].values()))
-        search_probs = number_of_node_visits/np.sum(number_of_node_visits)
+        # Consider Dirichlet noise like in alpha zero for exploration
+
+        values = self.state_values[fen_string].values()
+        moves = list(self.state_values[fen_string].keys())
+        n_values = np.array(list(zip(*values))[0])
+
+        # Get flattened probability distribution
+        n_values = torch.from_numpy(n_values)
+        n_values = n_values ** (1 / temperature)
+        search_probs = torch.nn.functional.softmax(n_values.float())
+        move = random.choices(moves, search_probs)[0]
 
         # Return list of uci_moves and corresponding search probabilities
-        return list(self.n_values[fen_string].keys()), search_probs
+        return list(self.state_values[fen_string].keys()), search_probs, move
 
- 
     def search(self, board, nnet):
-        """Method for performing a search on the tree.
-        """
+        """Descends on the search tree and expands unvisited states"""
         fen_string = board.fen()
+
         # Assigns reward if game is over
         if board.is_game_over():
             return 1 if board.outcome().winner else 0
 
-        # Checks to see if the node has been visited
+        # Checks to see if the node has been visited (expansion)
         if fen_string not in self.states_visited:
-            # Add the state to the list of visited nodes
             self.states_visited.append(fen_string)
 
-            # TODO: Get neural network state representation from fen_string
-            # input_state = get_cnn_input(fen_string)
-
-            # Return predictions and value from the nnet at the current state
-            value, policy = nnet.predict(board, None)
+            # Get predictions and value from the nnet at the current state
+            policy, value = nnet(get_cnn_input(board).float())
```

Comment:
Do we need this cast to float here? If it needs to be a float, I think that should be done within the definition of `get_cnn_input(board)` rather than here